sb3_contrib/trpo/trpo.py

```diff
@@ -0,0 +1,361 @@
+                        only_inputs=True,
+                    )
+                    if kl_param_grad is not None:
+                        g_grad, *_ = torch.autograd.grad(
+                            policy_obj, param, retain_graph=True, only_inputs=True
+                        )
+
+                        grad_shape.append(kl_param_grad.shape)
+                        grad_kl.append(kl_param_grad.view(-1))
+                        g.append(g_grad.view(-1))
+                        params.append(param)
+                    else:
+                        value_only_params.append(param)
+
+                g = torch.cat(g)
+                grad_kl = torch.cat(grad_kl)
+
+                def Hpv(v, retain_graph=True):
+                    jvp = (grad_kl * v).sum()
+                    return flat_grad(jvp, params, retain_graph=retain_graph) + self.cg_damping * v
+
+                s = cg_solver(Hpv, g, max_iter=self.cg_max_steps)
+
+                beta = 2 * self.target_kl
+                beta /= torch.matmul(s, Hpv(s, retain_graph=False))
+                beta = torch.sqrt(beta)
+
+                alpha = 1
+                orig_params = [param.detach().clone() for param in params]
+
+                line_search_success = False
+                with torch.no_grad():
+                    for i in range(self.ls_steps):
+
+                        j = 0
+                        for param, orig_param, shape in zip(params, orig_params, grad_shape):
+                            k = param.numel()
+                            param.data = orig_param.data + alpha * beta * s[j : (j + k)].view(shape)
+                            j += k
+
+                        _, log_prob, _ = self.policy.evaluate_actions(
+                            rollout_data.observations, actions
+                        )
+
+                        ratio = th.exp(log_prob - rollout_data.old_log_prob)
+                        new_policy_obj = (advantages * ratio).mean()
+
+                        kl_div = kl_divergence(
+                            distribution.distribution, old_distribution.distribution
+                        ).mean()
+
+                        if (kl_div < self.target_kl) and (new_policy_obj > policy_obj):
+                            line_search_success = True
+                            break
+
+                        alpha *= self.ls_alpha
+
+                    line_search_results.append(line_search_success)
+
+                    if not line_search_success:
+                        for param, orig_param in zip(params, orig_params):
+                            param.data = orig_param.data.clone()
+
+                        po_values.append(policy_obj.item())
+                        kl_divergences.append(0)
+                    else:
+                        po_values.append(new_policy_obj.item())
+                        kl_divergences.append(kl_div.item())
+
+                for _ in range(self.n_critic_updates):
```

Comment:
that would make more sense to do value update before policy update no?

Reply:
I understand, I think TRPO was not meant to have shared architecture.
I will probably do a PR in SB3 to allow completely independent actor/critic (currently the feature extractor is still shared)
in the meantime, we can probably try both and see if it affects results (for simple architecture, the actor/critic are independent)